src/main/java/seedu/address/model/book/BookStatus.java

```diff
@@ -0,0 +1,172 @@
+    }
+
+    /**
+     * A static function for creating an AVAILABLE book status.
+     *
+     * @return An AVAILABLE BookStatus object.
+     */
+    public static BookStatus createAvailableBookStatus() {
+        return new BookStatus(AVAILABLE, Optional.empty(), Optional.empty(),
+                Optional.empty());
+    }
+
+    /**
+     * Returns true if a given parameters can be converted to a valid book status.
+     */
+    public static boolean isValidStatus(BookStatusType bookStatusType, Optional<Patron> borrower,
+                                        Optional<String> borrowDate, Optional<String> returnDate) {
+        requireAllNonNull(bookStatusType, borrower, borrowDate, returnDate);
+
+        // A book with available status should not have any borrower, borrowing or returning dates
+        boolean isAvailableConstraintsSatisfied = bookStatusType == AVAILABLE
+                && borrower.isEmpty() && borrowDate.isEmpty() && returnDate.isEmpty();
+
+        // A book with borrowed status must have corresponding borrower, borrow and return dates.
+        String borrowDateString = borrowDate.orElse("");
+        String returnDateString = returnDate.orElse("");
+        boolean isValidBorrowRegex = borrowDateString.matches(VALIDATION_REGEX);
+        boolean isValidReturnRegex = returnDateString.matches(VALIDATION_REGEX);
+        boolean isBorrowedConstraintsSatisfied = bookStatusType == BORROWED
+                && borrower.isPresent() && borrowDate.isPresent()
+                && returnDate.isPresent() && isValidBorrowRegex && isValidReturnRegex
+                && isParseable(borrowDateString) && isParseable(returnDateString);
+
+        return isAvailableConstraintsSatisfied || isBorrowedConstraintsSatisfied;
+    }
+
+    /**
+     * Returns true if the status is AVAILABLE status.
+     */
+    public boolean isAvailable() {
+        return bookStatusType == AVAILABLE;
+    }
+
+    /**
+     * Returns true if the status is BORROWED status.
+     */
+    public boolean isBorrowed() {
+        return bookStatusType == BORROWED;
+    }
+
+    /**
+     * Returns a String representing the borrow date of this book, or an empty string if this book is not borrowed.
+     */
+    public String getBorrowDateString() {
+        if (borrowDate.isEmpty()) {
+            return "";
+        }
+        return STATUS_DATE_FORMAT.format(borrowDate.get());
+    }
+
+    /**
+     * Returns a String representing the return date of this book, or an empty string if this book is not borrowed.
+     */
+    public String getReturnDateString() {
+        if (returnDate.isEmpty()) {
+            return "";
+        }
+        return STATUS_DATE_FORMAT.format(returnDate.get());
+    }
+
+    /**
+     * Returns a String representing the borrower's name of this book, or an empty string if this book is not borrowed.
+     */
+    public String getBorrowerName() {
+        return borrower.map(patron -> patron.getName().toString()).orElse("");
+    }
```

Comment:
Good use of map and orElse to get the borrower name, making the code simple and understandable :100: